### PR TITLE
Don't allow duplicate file/folder entries to be added to sql projects

### LINKED ISSUE #14439: `addToProject` API adds duplicate items
- Azure Data Studio Version: 1.27.0-insider (user setup)
Commit: 9148aa1ed552cc89b7362f8132c2dec437ce5537

**Steps to Reproduce:**

1. Using new `addToProject` API from [this change](https://github.com/microsoft/azuredatastudio/pull/14391), call it twice with the same file:
```typescript
api.addToProject(vscode.Uri.file('Project2.sqlproj'), [vscode.Uri.file('Table2.sql')]);
api.addToProject(vscode.Uri.file('Project2.sqlproj'), [vscode.Uri.file('Table2.sql')]);
```
2.  `.sqlproj` gets two items with `Build` build action for the same file:
```xml
<ItemGroup>
  <Build Include="Table2.sql" />
  <Build Include="Table2.sql" />
</ItemGroup>
```

**Expected result:**
I think we should verify that files are not there yet when adding. I can't do this externally as there appears to be no API to list files that are currently in the project and just the presence of the file on disk does not mean it is included in the project.

### PR DESCRIPTION
**Pull request by kisantia**
This PR fixes #14439. This adds an error when a file or folder is attempted to be added to the project again.

### CODE REVIEW

**Review by SakshiS-harma — APPROVED** [review]

File: extensions/sql-database-projects/src/models/project.ts [hunk 1]
```diff
@@ -266,6 +266,11 @@ export class Project {
 	public async addFolderItem(relativeFolderPath: string): Promise<FileProjectEntry> {
 		const absoluteFolderPath = path.join(this.projectFolderPath, relativeFolderPath);
 
+		// check if folder already has been added to sqlproj
+		if (this.files.find(f => f.relativePath === relativeFolderPath)) {
```
Comment (alanrenmsft) [inline extensions/sql-database-projects/src/models/project.ts:270]: how about casing?

**Review by alanrenmsft — APPROVED** [review]

File: extensions/sql-database-projects/src/models/project.ts (hunk 1 above)
Comment (Benjin) [reply extensions/sql-database-projects/src/models/project.ts:270]: Windows and macOS are both case-insensitive, Linux isn't (by default)... It's probably fine to .toLowerCase() everything anyway, though.

**Review by Benjin — APPROVED** [review]

**Comment by nahk-ivanov** [comment]
Just got the notification that the issue was resolved...
Can we at least have a boolean flag to ignore duplicates, rather than throwing an error? As the issue stated:

> I can't do this externally as there appears to be no API to list files that are currently in the project...

So it seems like after this change goes out - we will be blocked as I have no means of filtering the list before feeding it to `addToProject`.

**Comment by kisantia** [comment]
This change won't be released until next month. Before that, I plan on exposing the other apis needed for getting the files in the project. I still need to do a little testing, but [these](https://github.com/microsoft/azuredatastudio/compare/kisantia/exposeProject?expand=1) are the changes so far(the changes in sqldbproj.d.ts are the relevant ones for your extension). Let me know if there's anything else you would need for filtering.

**Comment by nahk-ivanov** [comment]
Yeah, that should work, but I think we can update `addToProject(list: vscode.Uri[]): Promise<void>;` to return a collection of `IFileProjectEntry`. Then instead of throwing an error - it can dedupe internally and return already existing project entries for the duplicate files. Do you think that will be too much work?

I feel like adding files is a rather common functionality and it will be rare when we want an error for duplicates (the scenario is basically "I want to check if file is already in the project, so I try to add it and catch an exception"). What cases do you envision where having an error instead of ignoring the duplicates will be beneficial? Even when you get to "Add existing item..." functionality, you will probably want to ignore the duplicates, similarly to how Visual Studio does. At that point I think we would want to reuse the functionality, instead of duplicating it in multiple different extensions.

**Pull request merged**